src/Index/Index.php

```diff
@@ -216,10 +272,174 @@ public function unserialize($serialized)
     public function serialize()
     {
         return serialize([
-            'definitions' => $this->definitions,
+            'definitions' => iterator_to_array($this->getDefinitions(), true),
             'references' => $this->references,
             'complete' => $this->complete,
             'staticComplete' => $this->staticComplete
         ]);
     }
+
+    /**
+     * Returns a Genrerator containing all the into the given $storage recursively.
+     * The generator yields key => value pairs, eg
+     * 'Psr\Log\LoggerInterface->log()' => $definition
+     *
+     * @param array &$storage
+     * @param string $prefix (optional)
+     * @return \Generator
+     */
+    private function yieldDefinitionsRecursively(array &$storage, string $prefix = ''): \Generator
+    {
+        foreach ($storage as $key => $value) {
+            if (!is_array($value)) {
+                yield sprintf('%s%s', $prefix, $key) => $value;
+            } else {
+                yield from $this->yieldDefinitionsRecursively($value, sprintf('%s%s', $prefix, $key));
+            }
+        }
+    }
+
+    /**
+     * Splits the given FQN into an array, eg :
+     * 'Psr\Log\LoggerInterface->log' will be ['Psr', '\Log', '\LoggerInterface', '->log()']
+     * '\Exception->getMessage()'     will be ['\Exception', '->getMessage()']
+     * 'PHP_VERSION'                  will be ['PHP_VERSION']
+     *
+     * @param string $fqn
+     * @return array
+     */
+    private function splitFqn(string $fqn): array
```

Comment:
Wouldn't this be much simpler with
```php
preg_split("/(\\\\|::|->)/", $fqn, 0, PREG_SPLIT_DELIM_CAPTURE);
```
?

Reply:
usally the string functions are way faster then preg_*
you would want to measure it I guess

Reply:
Hm, I find it hard to believe that all those `array_slice`, `array_pop` calls etc are faster than a single `preg_split`. Maybe you could commit the code you used to generate the benchmark in the description?